### PR TITLE
Add v1alpha2 MPIJob controller

### PR DESCRIPTION
**Pull request by terrytangyuan**
This is the new MPIJob controller based on the API design in https://github.com/kubeflow/mpi-operator/pull/95. Still a work-in-progress but creating a PR here for any immediate feedback.

Note that I created a new folder named "controllers" that includes both `v1alpha1` and `v1alpha2` MPIJob controllers.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/mpi-operator/98)

### CODE REVIEW

File: pkg/controllers/v1alpha2/mpi_job_controller.go [hunk 1]
```diff
@@ -0,0 +1,1158 @@
...
+		}
+
+		// Get the launcher RoleBinding for this MPIJob.
+		if rb, err := c.getLauncherRoleBinding(mpiJob); rb == nil || err != nil {
+			return err
+		}
+
+		// Get the PDB for this MPIJob
+		if c.enableGangScheduling {
+			if pdb, err := c.getOrCreatePDB(mpiJob, workerReplicas); pdb == nil || err != nil {
```
Comment (fisherxu) [inline pkg/controllers/v1alpha2/mpi_job_controller.go]: should move to podgroup here :) pdb has deprecated in kube-batch

Comment (terrytangyuan) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: For sure yes but it's probably better done in a separate PR. Contribution is welcomed!

**Comment by gaocegege** [comment]
Feel free to @ me when the PR is ready to review.

File: pkg/controllers/v1alpha2/mpi_job_controller.go (hunk 1 above)
Comment (fisherxu) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: Okey,so we should push this pr merged first

File: pkg/controllers/v1alpha2/mpi_job_controller.go [hunk 2]
```diff
@@ -0,0 +1,1158 @@
...
+			Name:  "OMPI_MCA_plm_rsh_agent",
+			Value: fmt.Sprintf("%s/%s", configMountPath, kubexecScriptName),
+		},
+		corev1.EnvVar{
+			Name:  "OMPI_MCA_orte_default_hostfile",
+			Value: fmt.Sprintf("%s/%s", configMountPath, hostfileName),
+		})
+	// Not assign any resource limits and requests to launcher pod
+	container.Resources.Limits = nil
+	container.Resources.Requests = nil
```
Comment (fisherxu) [inline pkg/controllers/v1alpha2/mpi_job_controller.go]: the resource maybe can read from job launcher spec, user can specify resource to launcher pod if needed

Comment (terrytangyuan) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: Yes, this was work-in-progress so I haven't pushed all the changes yet. Just updated the PR.

**Comment by terrytangyuan** [comment]
This PR is now ready for review. 

Note that this controller is not finished yet as some tests are still failing but I do not have dedicated time for this anytime soon. I see continuous contributions from the community to v1alpha1 controller but eventually require more work to land in v1alpha2 controller so I want to wrap up this PR quickly so new contributions could land in new version of the controller in the meantime. The community is welcome to help fixing the tests and I will certainly look into more if I have time. 

This is based on an older version ever since https://github.com/kubeflow/mpi-operator/pull/95 so there are some changes that are not in this new controller yet (which could be some follow-up work by the community but keeping them here for reference later): https://github.com/kubeflow/mpi-operator/pull/106 https://github.com/kubeflow/mpi-operator/pull/105 https://github.com/kubeflow/mpi-operator/pull/100.

/cc @gaocegege @rongou @madhukarkm @Nivedita-V

File: pkg/controllers/v1alpha2/mpi_job_controller.go (hunk 1 above)
Comment (terrytangyuan) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: Just updated the PR and commented. Feel free to help continue working on this if you are interested.

File: .travis.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ before_install:
 - dep ensure
 
 install:
-- go get -t -d ./...
+- go get -t -d ./pkg/controllers/v1alpha1/...
```
Comment (rongou) [inline .travis.yml]: This just builds the code, right? The new controller should at least compile. :)

File: .travis.yml [hunk 4]
```diff
@@ -9,7 +9,7 @@ before_install:
...
 
 script:
-- go test ./...
+- go test ./pkg/controllers/v1alpha1/...
```
Comment (rongou) [inline .travis.yml]: Probably do something like
`go test $(go list ./... | grep -v /v1alpha2/)`

File: .travis.yml (hunk 3 above)
Comment (terrytangyuan) [reply .travis.yml]: Good catch! Just fixed it.

File: .travis.yml (hunk 4 above)
Comment (terrytangyuan) [reply .travis.yml]: Just updated to use `go test ./pkg/controllers/v1alpha1/... ./cmd/...` instead, excluding `v1alpha2` test execution for now.

Comment (terrytangyuan) [reply .travis.yml]: Actually I just realized there isn’t any test in cmd/, I just changed it back to “go test ./pkg/controllers/v1alpha1/...”? Are there any other tests that need to be run I may have missed?

File: pkg/controllers/v1alpha2/mpi_job_controller.go (hunk 1 above)
Comment (fisherxu) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: sure, I will interested working on this :)

File: .travis.yml (hunk 4 above)
Comment (rongou) [reply .travis.yml]: My point is we should explicitly exclude the v1alpha2 tests for now, so we don't forget to run other tests if/when we add them in the future.

Comment (terrytangyuan) [reply .travis.yml]: Got it. Done.

**Comment by terrytangyuan** [comment]
Thanks everyone! 
/approve

**Pull request merged**